File: app/src/main/java/me/instabattle/app/activities/CreateBattleActivity.java
package me.instabattle.app.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.gms.maps.model.LatLng;

import me.instabattle.app.R;
import me.instabattle.app.images.Util;
import me.instabattle.app.managers.BattleManager;
import me.instabattle.app.managers.PhotoManager;
import me.instabattle.app.models.Battle;
import me.instabattle.app.models.Entry;
import me.instabattle.app.services.LocationService;
import me.instabattle.app.settings.State;
import me.instabattle.app.utils.Utils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class CreateBattleActivity extends Activity {

    private static final String TAG = "CreateBattleActivity";

    private static final int BATTLE_TITLE_MINIMUM_LENGTH = 1;

    private TextView newBattleTitle;
    private TextView newBattleDescription;
    private TextView newBattleRadius;
    private ImageView newBattlePhoto;

    private static String savedTitle = "";
    private static String savedDescription = "";
    private static String savedRadius = "";

    private String getBattleTitle() {
        return newBattleTitle.getText().toString();
    }

    private String getBattleDescription() {
        return newBattleDescription.getText().toString();
    }

    private String getBattleRadius() {
        return newBattleRadius.getText().toString();
    }

    public static byte[] photoBytes;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_battle);

        newBattleTitle = (TextView) findViewById(R.id.newBattleTitle);
        newBattleDescription = (TextView) findViewById(R.id.newBattleDescription);
        newBattleRadius = (TextView) findViewById(R.id.newBattleRadius);
        newBattlePhoto = (ImageView) findViewById(R.id.newBattlePhoto);

        if (photoBytes != null) {
            newBattlePhoto.setImageBitmap(Util.decodeSampledBitmapFromBytes(photoBytes, 256, 256));
        }
    }

    @Override
    public void onResume() {
        super.onResume();

        newBattleTitle.setText(savedTitle);
        newBattleDescription.setText(savedDescription);
        newBattleRadius.setText(savedRadius);
    }

    private boolean validateBattle() {
        if (getBattleTitle().length() < BATTLE_TITLE_MINIMUM_LENGTH) {
            Utils.showToast(this, "Battle name should have at least " +
                    BATTLE_TITLE_MINIMUM_LENGTH + " characters!");
            return false;
        }
        if (photoBytes == null) {
            Utils.showToast(this, "You should take a first photo in battle!");
            return false;
        }
        return true;
    }

    public void sendBattle(View v) {
        if (!validateBattle()) {
            return;
        }

        final LatLng loc = LocationService.getCurrentLocation();
        BattleManager.createAndDo(State.currentUser.getId(),
                getBattleTitle(),
                loc.latitude,
                loc.longitude,
                getBattleDescription(),
                Integer.parseInt(getBattleRadius()),
                new Callback<Battle>() {
                    @Override
                    public void onResponse(Call<Battle> call, Response<Battle> response) {
                        Log.d(TAG, "new battle was sent");
                        addFirstEntry(response.body());
                    }

                    @Override
                    public void onFailure(Call<Battle> call, Throwable t) {
                        Utils.showToast(CreateBattleActivity.this, "Failed to send battle to server, try again later.");
                        Log.e(TAG, "can't send new battle: " + t);
                    }
        });

        State.creatingBattle = false;
        Intent goToMap = new Intent(this, MapActivity.class);

        clearFields();
        startActivity(goToMap);
    }

    private void addFirstEntry(final Battle battle) {
        battle.createEntryAndDo(new Callback<Entry>() {
            @Override
            public void onResponse(Call<Entry> call, Response<Entry> response) {
                PhotoManager.upload(response.body().getImageName(), photoBytes);
                Log.d(TAG, "new entry was sent");
            }

            @Override
            public void onFailure(Call<Entry> call, Throwable t) {
                addFirstEntry(battle);
                Log.e(TAG, "can't send first entry: " + t);
            }
        });
    }

    public void takeBattlePhoto(View v) {
        ParticipatingActivity.gotHereFrom = CreateBattleActivity.class;
        Intent takePhoto = new Intent(this, ParticipatingActivity.class);

        saveFields();
        startActivity(takePhoto);
    }

    private void saveFields() {
        savedTitle = getBattleTitle();
        savedDescription = getBattleDescription();
        savedRadius = getBattleRadius();
    }

    private void clearFields() {
        savedTitle = savedDescription = savedRadius = "";
    }

    @Override
    public void onBackPressed() {
        Intent intent = new Intent(this, MapActivity.class);

        clearFields();
        State.creatingBattle = false;
        startActivity(intent);
    }
}

File: app/src/main/java/me/instabattle/app/activities/CreateBattleActivity.kt
package me.instabattle.app.activities

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import android.widget.TextView

import me.instabattle.app.R
import me.instabattle.app.images.Util
import me.instabattle.app.managers.BattleManager
import me.instabattle.app.managers.PhotoManager
import me.instabattle.app.models.Battle
import me.instabattle.app.models.Entry
import me.instabattle.app.services.LocationService
import me.instabattle.app.settings.State
import org.jetbrains.anko.info
import org.jetbrains.anko.error
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.toast
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class CreateBattleActivity : DefaultActivity() {
    private lateinit var newBattleTitle: TextView
    private lateinit var newBattleDescription: TextView
    private lateinit var newBattleRadius: TextView
    private lateinit var newBattlePhoto: ImageView

    private val battleTitle: String
        get() = newBattleTitle.text.toString()

    private val battleDescription: String
        get() = newBattleDescription.text.toString()

    private val battleRadius: String
        get() = newBattleRadius.text.toString()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_create_battle)

        newBattleTitle = findViewById(R.id.newBattleTitle)
        newBattleDescription = findViewById(R.id.newBattleDescription)
        newBattleRadius = findViewById(R.id.newBattleRadius)
        newBattlePhoto = findViewById(R.id.newBattlePhoto)

        if (photoBytes != null) {
            newBattlePhoto.setImageBitmap(Util.decodeSampledBitmapFromBytes(photoBytes, 256, 256))
        }
    }

    public override fun onResume() {
        super.onResume()

        newBattleTitle.text = savedTitle
        newBattleDescription.text = savedDescription
        newBattleRadius.text = savedRadius
    }

    private fun validateBattle(): Boolean {
        if (battleTitle.length < BATTLE_TITLE_MINIMUM_LENGTH) {
            toast("Battle name should have at least  $BATTLE_TITLE_MINIMUM_LENGTH characters!")
            return false
        }
        if (photoBytes == null) {
            toast("You should take a first photo in battle!")
            return false
        }
        return true
    }

    fun sendBattle(v: View) {
        if (!validateBattle()) {
            return
        }

        val loc = LocationService.getCurrentLocation()
        BattleManager.createAndDo(State.currentUser.id,
                battleTitle,
                loc.latitude,
                loc.longitude,
                battleDescription,
                Integer.parseInt(battleRadius),
                object : Callback<Battle> {
                    override fun onResponse(call: Call<Battle>, response: Response<Battle>) {
                        info {"New battle was sent"}
                        if (response.isSuccessful) {
                            State.chosenBattle = response.body()
                            addFirstEntry(State.chosenBattle)
                            State.creatingBattle = false
                            val viewBattle = Intent(this@CreateBattleActivity, BattleActivity::class.java)
                            clearFields()
                            BattleActivity.gotHereFrom = MapActivity::class.java
                            startActivity(viewBattle)
                        } else {
                            info {"Couldn't create battle"}
                        }
                    }

                    override fun onFailure(call: Call<Battle>, t: Throwable) {
                        toast("Failed to send battle to server, try again later.")
                        error {"can't send new battle: $t"}
                    }
                })
    }

    private fun addFirstEntry(battle: Battle) {
        battle.createEntryAndDo(object : Callback<Entry> {
            override fun onResponse(call: Call<Entry>, response: Response<Entry>) {
                PhotoManager.upload(response.body()!!.imageName, photoBytes)
                 info {"new entry was sent"}
            }

            override fun onFailure(call: Call<Entry>, t: Throwable) {
                addFirstEntry(battle)
                error {"can't send first entry: $t"}
            }
        })
    }

    fun takeBattlePhoto(v: View) {
        ParticipatingActivity.gotHereFrom = CreateBattleActivity::class.java
        val takePhoto = Intent(this, ParticipatingActivity::class.java)

        saveFields()
        startActivity(takePhoto)
    }

    private fun saveFields() {
        savedTitle = battleTitle
        savedDescription = battleDescription
        savedRadius = battleRadius
    }

    private fun clearFields() {
        savedRadius = ""
        savedDescription = ""
        savedTitle = ""
    }

    override fun onBackPressed() {
        clearFields()
        State.creatingBattle = false
        startActivity<MapActivity>()
    }

    companion object {
        private const val BATTLE_TITLE_MINIMUM_LENGTH = 1

        private var savedTitle = ""
        private var savedDescription = ""
        private var savedRadius = ""

        var photoBytes: ByteArray? = null
    }
}